mllib/src/main/scala/org/apache/spark/ml/classification/FMClassifier.scala

```diff
@@ -235,6 +236,13 @@ class FMClassifier @Since("3.0.0") (
     model.setSummary(Some(summary))
   }
 
+  override def estimateModelSize(dataset: Dataset[_]): Long = {
+    val numFeatures = DatasetUtils.getNumFeatures(dataset, $(featuresCol))
```

Comment:
I am worried that executing the input query multiple times will be considered wasteful. I am wondering if we should do the check while fitting instead; we fail as soon as the model gets too large.

Reply:
DatasetUtils.getNumFeatures is quite cheap, it will try to fetch `numFeatures` from the metadata, and if there is no such metadata, it just infer the `numFeatures` from the first row.